### PR TITLE
Tests for config validators

### PR DESCRIPTION
**Pull request by barnettwilliam**
Tests for ToolConfigValidator and ActivityConfigValidator.

Code coverage for the modules 100% though further tests can be added to check the contents of each error message against the error raised.

### CODE REVIEW

File: platform/test/spec/testActivityConfigValidatorSpec.js [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+            let e = errors[0];
+            
+            // Check the expected results
+            expect(errors).toHaveSize(1);
+            checkErrorPopulated(e);
+        })
+    }) 
+})
+
+function checkErrorPopulated(error){
```
Comment (szschaler) [inline platform/test/spec/testActivityConfigValidatorSpec.js:77]: These checks should be more specific and test that the correct errors are reported.

File: platform/test/spec/testActivityConfigValidatorSpec.js [hunk 2]
```diff
@@ -0,0 +1,84 @@
+/*global describe, it, expect, beforeEach --  functions provided by Jasmine */
+import { ActivityConfigValidator } from "../../src/ActivityConfigValidator.js"
+import { ACTIVITY_2PANELS_1ACTION } from "../resources/TestActivityFiles.js";
+import { ConfigValidationError } from "../../src/ConfigValidationError.js"
+
+
+const EXPECTED_FILE_TYPE = "ActivityConfig"; 
+
+describe("ActivityConfigValidator", () => {
```
Comment (szschaler) [inline platform/test/spec/testActivityConfigValidatorSpec.js:9]: Is this really testing all error conditions that the configuration validator checks for?

File: platform/test/spec/testToolConfigValidatorSpec.js [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        it("can be created", () => {
+            // Call the target object
+            let tcv = new ToolConfigValidator();
+            
+            // Check the expected results
+            expect(tcv).toBeInstanceOf(ToolConfigValidator);
+        })
+    })
+
+    describe("validate tool configration", () => {
```
Comment (szschaler) [inline platform/test/spec/testToolConfigValidatorSpec.js:21]:
```suggestion
    describe("validate tool configuration", () => {
```

File: platform/test/spec/testToolConfigValidatorSpec.js [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+            // Call the target object
+            let errors = tcv.validateConfigFile(toolConfig);
+            let e = errors[0];
+            
+            // Check the expected results
+            expect(e).toBeInstanceOf(ConfigValidationError);
+            expect(e.fileType).toEqual(EXPECTED_FILE_TYPE);
+        })
+
+        it("returns an error if the config has no id key", () => {
```
Comment (szschaler) [inline platform/test/spec/testToolConfigValidatorSpec.js:51]: Merge with previous test.

File: platform/test/spec/testToolConfigValidatorSpec.js [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+            let e = errors[0];
+            
+            // Check the expected results
+            expect(errors).toHaveSize(1);
+            checkErrorPopulated(e);
+        })
+    }) 
+})
+
+function checkErrorPopulated(error){
```
Comment (szschaler) [inline platform/test/spec/testToolConfigValidatorSpec.js:77]: Can we check more specifically the error we expect to see?

File: platform/test/spec/testToolConfigValidatorSpec.js [hunk 6]
```diff
@@ -0,0 +1,84 @@
+/*global describe, it, expect, beforeEach --  functions provided by Jasmine */
+import { ToolConfigValidator } from "../../src/ToolConfigValidator"
+import { TOOL_1PANELDEF_1FUNCTION } from "../resources/TestToolFiles.js";
+import { ConfigValidationError } from "../../src/ConfigValidationError.js"
+
+
+const EXPECTED_FILE_TYPE = "ToolConfig"; 
+
+describe("ToolConfigValidator", () => {
```
Comment (szschaler) [inline platform/test/spec/testToolConfigValidatorSpec.js:9]: Is this really testing all error conditions that the validator checks for?

**Review by szschaler — CHANGES_REQUESTED** [review]

File: platform/test/spec/testActivityConfigValidatorSpec.js (hunk 2 above)
Comment (barnettwilliam) [reply platform/test/spec/testActivityConfigValidatorSpec.js:8]: No its schema validation, there is another validator ActivityValidator that does more of the semantic checking that has its own tests in #190.

File: platform/test/spec/testToolConfigValidatorSpec.js (hunk 6 above)
Comment (barnettwilliam) [reply platform/test/spec/testToolConfigValidatorSpec.js:8]: No its schema validation, there is another validator ActivityValidator that does more of the semantic checking that has its own tests in #190.

File: platform/test/spec/testActivityConfigValidatorSpec.js (hunk 1 above)
Comment (barnettwilliam) [reply platform/test/spec/testActivityConfigValidatorSpec.js]: Improved the `CheckErrorPopulated()` function to test each attribute of the error and updated tests.  Fixed in cb677e9410c8154ce39faa02c77f690e60f89964.

File: platform/test/spec/testToolConfigValidatorSpec.js (hunk 5 above)
Comment (barnettwilliam) [reply platform/test/spec/testToolConfigValidatorSpec.js]: Improved the `CheckErrorPopulated()` function to test each attribute of the error and updated tests.  Fixed in cb677e9410c8154ce39faa02c77f690e60f89964.

File: platform/test/spec/testActivityConfigValidatorSpec.js [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+            // Call the target object
+            let errors = acv.validateConfigFile(activityConfig);
+            let e = errors[0];
+            
+            // Check the expected results
+            expect(e).toBeInstanceOf(ConfigValidationError);
+            expect(e.fileType).toEqual(EXPECTED_FILE_TYPE);
+        })
+
+        it("returns an error if the config has no id key", () => {
```
[the remark this answers is not in the document]
Comment (barnettwilliam) [reply platform/test/spec/testActivityConfigValidatorSpec.js]: This is being check by `CheckErrorPopulated()` so removed. Fixed in cb677e9410c8154ce39faa02c77f690e60f89964

File: platform/test/spec/testToolConfigValidatorSpec.js (hunk 4 above)
Comment (barnettwilliam) [reply platform/test/spec/testToolConfigValidatorSpec.js]: This is being check by `CheckErrorPopulated()` so removed. Fixed in cb677e9410c8154ce39faa02c77f690e60f89964

File: platform/test/resources/TestUtility.js [hunk 8]
```diff
@@ -11,3 +14,57 @@ export function configObjectEquals(configA, configB){
 
     return (strConfigA === strConfigB)
 }
+
+// Jasmine custom matchers
+export const customMatchers = {
+    
+    /**
+     * Matcher to check the expected keywords are present in the give input.
```
Comment (szschaler) [inline platform/test/resources/TestUtility.js:22]:
```suggestion
     * Matcher to check the expected keywords are present in the given input.
```

File: platform/test/resources/TestUtility.js [hunk 9]
```diff
@@ -11,3 +14,57 @@ export function configObjectEquals(configA, configB){
...
+     * where actual is a Sting and expected is an array of strings that contains
```
Comment (szschaler) [inline platform/test/resources/TestUtility.js:23]:
```suggestion
     * where actual is a String and expected is an array of strings that contains 
```

File: platform/test/resources/TestUtility.js [hunk 10]
```diff
@@ -11,3 +14,57 @@ export function configObjectEquals(configA, configB){
...
+    toContainKeywords : function(matchersUtil){
+        return {
+            compare: function(actual, expected){
+
+                const result = {
+                };
+
+                // Check the condition
+                let containedResults =  expected.map( (akeyword)=> matchersUtil.contains(actual, akeyword) )
+                result.pass = containedResults.every(r => r);
```
Comment (szschaler) [inline platform/test/resources/TestUtility.js:35]: How about

```suggestion
                result.pass = expected.every( (akeyword)=> matchersUtil.contains(actual, akeyword) );
```

as a simplification?

File: platform/test/resources/TestUtility.js [hunk 11]
```diff
@@ -11,3 +14,57 @@ export function configObjectEquals(configA, configB){
...
+
+// Jasmine custom matchers
+export const customMatchers = {
+    
+    /**
+     * Matcher to check the expected keywords are present in the give input.
+     * where actual is a Sting and expected is an array of strings that contains 
+     * the keyword to check.
+     */
+    toContainKeywords : function(matchersUtil){
```
Comment (szschaler) [inline platform/test/resources/TestUtility.js:26]:
```suggestion
    toContainAllKeywords : function(matchersUtil){
```

File: platform/test/resources/TestUtility.js [hunk 12]
```diff
@@ -11,3 +14,57 @@ export function configObjectEquals(configA, configB){
...
+    const MIN_LENGTH = 3;
+    expect(error).toBeInstanceOf(ConfigValidationError);
+    
+    expect(error.category.length).toBeGreaterThanOrEqual(MIN_LENGTH);
+    expect(error.message.length).toBeGreaterThanOrEqual(MIN_LENGTH);
+    expect(error.location.length).toBeGreaterThanOrEqual(MIN_LENGTH);
+   
+    expect(error.category).toEqual(category);
+    expect(error.fileType).toEqual(type);
+    expect(error.message).toContainKeywords(messageKeywords);
```
Comment (szschaler) [inline platform/test/resources/TestUtility.js:68]:
```suggestion
    expect(error.message).toContainAllKeywords(messageKeywords);
```

**Review by szschaler — APPROVED** [review]
LGTM, subject to some small suggestions

**Comment by barnettwilliam** [comment]
Thank you nice suggestions, fixed in e22ecd6a2187f3d7d79a6e726f53601a042c193e

**Pull request merged**